cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -0,0 +1,410 @@
+      }
+      List<String> reportRecords = new ArrayList<>();
+      long lineCount = 0;
+      Location reportDir = reportIdDir.append(REPORT_DIR);
+      Location reportFile = null;
+      // TODO: assume only one report file for now
+      for (Location file : reportDir.list()) {
+        if (file.getName().endsWith(".json")) {
+          reportFile = file;
+          break;
+        }
+      }
+      if (reportFile == null) {
+        responder.sendError(500, "No files found for report " + reportId);
+        return;
+      }
+      // TODO: use cache to store content of the reports
+      // Read the report file and add lines starting from the position of offset to the result until the result reaches
+      // the limit
+      try (BufferedReader br = new BufferedReader(new InputStreamReader(reportFile.getInputStream()))) {
+        String line;
+        while ((line = br.readLine()) != null) {
+          // skip lines before the offset
+          if (lineCount++ < offset) {
+            continue;
+          }
+          if (reportRecords.size() == limit) {
+            break;
+          }
+          reportRecords.add(line);
+        }
+      }
+      // Get the total number of records from the COUNT file
+      String total =
+        new String(ByteStreams.toByteArray(reportIdDir.append(COUNT_FILE).getInputStream()), Charsets.UTF_8);
+      responder.sendJson(200, new ReportContent(offset, limit, Integer.parseInt(total), reportRecords));
+    }
+
+    @POST
+    @Path("/reports")
+    public void executeReportGeneration(HttpServiceRequest request, HttpServiceResponder responder)
+      throws IOException {
+      String requestJson = Charsets.UTF_8.decode(request.getContent()).toString();
+      LOG.debug("Received report generation request {}", requestJson);
+      ReportGenerationRequest reportRequest;
+      try {
+        reportRequest = decodeRequestBody(requestJson, REPORT_GENERATION_REQUEST_TYPE);
+        reportRequest.validate();
+      } catch (IllegalArgumentException e) {
+        responder.sendError(400, e.getMessage());
+        return;
+      }
+
+      String reportId = ReportIds.generate().toString();
+      Location reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      if (!reportIdDir.mkdirs()) {
+        responder.sendError(500, "Failed to create a new directory for report " + reportId);
+        return;
+      }
+      LOG.debug("Created report base directory {} for report {}", reportIdDir, reportId);
+      // Create a _START file to indicate the start of report generation
+      Location startFile = reportIdDir.append(START_FILE);
+      if (!startFile.createNew()) {
+        reportIdDir.delete();
+        responder.sendError(500, "Failed to create a _START file for report " + reportId);
+        return;
+      }
+      // Save the report generation request in the _START file
+      try (PrintWriter writer = new PrintWriter(startFile.getOutputStream())) {
```

Comment:
Again, alway provide `Charset` when you create a `Writer` from an `OutputStream`.